### ISSUE TITLE
Additional keyboard and keybind support

### ISSUE THREAD

**Issue by FatalError42O** [minetest/minetest#12488]
## Problem

It's extremely frustrating to create a control scheme that allows the player to do a number of different things when we are limited to only a few keys for controls. There's an express need for an additional expansion to the PC and desktop control scheme and possibly mobile.

currently, we're limited to about 3 different keys (if you're generous) for non-movement (or other hard-coded action) controls. This is frustrating as if you want to make a special menu, a action key (like a special ability), or a movement key (like crouch, prone), this is very limiting, and especially considering shift and zoom already having an action. While you can be creative, this basically still limits you to one key for any given action, and since most servers use aux for sprint... you basically have nothing. 

## Solutions

**PC/DESKTOP**

I propose two possible solutions (for PC)

**Solution A**: allow games to take matters into their own hands.
Allow mods to see the control binds of players, and see (almost) all pressed keys

_this would be used to get key binds, this is so mods can plan around existing player key binds, and even create their own binding system_
`minetest.get_player_key_binds(player) = {forward = "w", left = "a" . . . etc}`

_this would be a list of active keys, if a key is active it should not be on the list. This list would exclude some keys, like windows button (and linux equivalent), the function keys,  or anything else sensitive or not needed._
`minetest.get_player_held_controls(player) = {w = true, a = true . . . etc}`

these functions could alternatively be methods for the player object.

**Solution B**: allow game/world key binds
This would allow mods to create their own custom key bind settings for each player, and allow the client to interface with it.

on joining, the player would receive a list of key binds, this would be directly interfaceable with the "change keys" button, this should be modifiable by the server in post. The client would store these key binds by server in a human readable file format. This is especially important as to allow players to copy across versions etc, these files also should never delete themselves.

_self explanitory:_
`minetest.set_world_keybinds({keyname="key" . . . etc})`
`minetest.add_world_keybind(keyname, key)`
`minetest.on_player_change_keybinds(function(player, key, value))

**MOBILE**

I propose one solution for mobile, this would be the addition of "buttons" to the HUD definition. 
It's fairly simple actually:
the size of the button would be the same size as the image or text, just a box. The change would include this addition to the HUD definition:
`button = function(player)`
`button_image = "image.png"`
"button" would be called when the player presses down on it, when it's being pressed down upon, the image would change to button_image (if present). HOWEVER this system would require atleast 1 other PR(s)/issues to be put in place:
see relevant PRs:

#12264 user input mechanisms (for detecting touchpad) <---- most important, could be essential
#12367 window/screen size information (for better scaling)

If the inclusion of PR 12264 can not be made, HUD images with the button function should not be displayed on devices without capabilities to use them.
##additional

**Mobile should be delayed for obvious reasons**
The engine does not currently have the framework to allow for the support of mobile in regards to controls... but this doesn't mean we should limit ourselves and not make improvements where we can. 
We should improve PC control inputs and settings, and upgrade mobile options later, it's silly to limit ourselves to "X platform can't do Y so we should leave out this entirely basic feature for Z until we rework the entire engine".
Besides, games/mods that require additional controls are likely to hard to play on mobile, though I agree that's not our choice to make- it's the truth, and ignoring it is a massive oversight. 
**to reiterate**
Of course this doesn't mean we shouldn`t add support for mobile, but why do we need to ruin the fun for PC players who want to make something special.

(once again) relevant PRs 

#12264 user input mechanisms (for detecting touchpad) <---- most important, could be essential
#12367 window/screen size information (for better scaling)

**Comment by appgurueu** [comment]
On `minetest.get_player_held_controls(player)`: It's apparently pretty hard to ensure that key names are consistent across platforms. I could *still* see this being useful if games / servers provided their own "configure your keybinds" dialogs though - when the user is forced to configure their keybinds, it's irrelevant which names are used as long as they are unique. Only "default" keybinds might fail; and for trivially named keys (e.g. ASCII letters) ensuring consistent key names shouldn't be too hard.

That said, this would have to be fleshed out: What if the player is typing something into chat - should it still send keybinds? What about formspecs (relevant: https://github.com/minetest/minetest/pull/8679)?

**Comment by FatalError42O** [comment]
I think that for typing into chat, or any menu/formspec stuff it shouldn't, because there's currently no way to tell...

**Comment by FatalError42O** [comment]
if there was a way to tell, then yes, I'd say that it'd be a good idea.

**Comment by Montandalar** [comment]
For mobile, perhaps gesture support could be helpful?

**Comment by FatalError42O** [comment]
> For mobile, perhaps gesture support could be helpful?

This is entirely unnecessary 
Making a system for this would be complicated, and it serves no relevant or valid purpose.

**Comment by Montandalar** [comment]
A lazy post gets a lazy response. But you may have misunderstood what I mean by gestures. Gestures are the only way to interact with a phone that doesn't have external input devices like controllers or keyboards. Gestures include: tap, double tap, long tap, drag, pinch, pinch and rotate, swipe & multi-finger swipe, shake and others.

Double tap, and long tap are already bound to the placing and breaking actions, and dragging is bound to camera controls. A hot zone is used to emulate a kind of D-pad for movement. The mobile native way to enable extra actions is to use gestures. I would say at least let mod authors bind actions to multi finger swipe (left/right/up/down perhaps), pinch and shake.

Slightly off topic but here's an even more radical idea - maybe it's already possible? For vehicle controls on mobile, use the dpad for steering only and let the sneak/jump buttons be the brake and accelerator. Driving a mesecar or advtrains locomotive is pretty unintuitive with the dpad. If I try to turn too hard I brake, if I accidentally swipe a little bit left the train doors open...

My main point here is any decision we make about the engine can't simply ignore mobile. Yours does, and I don't think it holds merit for that reason. If I add the doodad key to PC controls so my mod can frobniscate my wim-wam, why am I going to have to explain to mobile users that "oh yeah, I know frobnostication would perfectly make sense as a shake gesture, for now it's just a button in that menu full of other buttons that do stuff"?

You'll have counterarguments, that's OK. But please don't disregard what I have to say out of hand again.

**Comment by FatalError42O** [comment]
> A lazy post gets a lazy response. But you may have misunderstood what I mean by gestures. Gestures are the only way to interact with a phone that doesn't have external input devices like controllers or keyboards. Gestures include: tap, double tap, long tap, drag, pinch, pinch and rotate, swipe & multi-finger swipe, shake and others.
> 
> Double tap, and long tap are already bound to the placing and breaking actions, and dragging is bound to camera controls. A hot zone is used to emulate a kind of D-pad for movement. The mobile native way to enable extra actions is to use gestures. I would say at least let mod authors bind actions to multi finger swipe (left/right/up/down perhaps), pinch and shake.
> 
> Slightly off topic but here's an even more radical idea - maybe it's already possible? For vehicle controls on mobile, use the dpad for steering only and let the sneak/jump buttons be the brake and accelerator. Driving a mesecar or advtrains locomotive is pretty unintuitive with the dpad. If I try to turn too hard I brake, if I accidentally swipe a little bit left the train doors open...
> 
> My main point here is any decision we make about the engine can't simply ignore mobile. Yours does, and I don't think it holds merit for that reason. If I add the doodad key to PC controls so my mod can frobniscate my wim-wams, why am I going to have to explain to mobile users that "oh yeah, I know frobnostication would perfectly make sense as a shake gesture, for now it's just a button in that menu full of other buttons that do stuff"?
> 
> You'll have counterarguments, that's OK. But please don't disregard what I have to say out of hand again.

It by no means ignores mobile, it implements the _equivalent_ system, there's no possibility to have the solution work the same for mobile, because they're different platforms with different input methods. Additionally, you still have not explained the use cases for gestures outside of vehicles, but even that use doesn't make sense.  Either way, this isn't the point.

The goal is to establish basic controls, I'm not saying it's necessarily a bad idea to add support for gestures, but it's something that should be at a later date, we all know that if there's a overly complicated suggestion or feature request, it'll take forever to get done or just get closed. It's more important to actually implement the BASICS. Also, a lot of the time, feature packed games like this that cannot practically use a menu for controls like this, aren't easy or practical for a mobile player to use anyway; this is why this issue is designed to primarily focus on the PC side of things, but still provides a viable equivalent solution to mobile.

>If I add the doodad key to PC controls so my mod can frobniscate my wim-wams, why am I going to have to explain to mobile users that "oh yeah, I know frobnostication would perfectly make sense as a shake gesture, for now it's just a button in that menu full of other buttons that do stuff"?

nobody said anything about a menu, if it was about menus we already HAVE menus, it's called FORMSPEC. The HUD buttons would allow you to create a button anywhere on the screen that modders could use for controls. Whatever a "shake gesture" could do (which is a feature even most dedicated mobile games don't use) could be easily accomplished, and even possibly more practical as a button.

TL:DR:
it's not practical, logical, or frankly even viable to add gestures at this time.

**Comment by Montandalar** [comment]
Okay, let's say we can ignore gestures at the moment. What makes interactive tappable HUDs really a mobile-only feature? Plenty of desktop users want those too. Having formspecs be completely modal sucks, you can't see what's going on in the world behind the formspec.

Many games let you release the mouse from camera controls and interact with smaller UI elements. For instance in Garry's Mod, the spawn menu does not disable movement controls, but it redirects the mouse from camera controls to a cursor; the leaderboard can also be used with the mouse to mute players in voice chat. In Dark Souls, you can navigate the menus while continuing to move your character; some keys are bound to the menu, while the WASD movement keys continue to be useable, and the mouse can be used as a cursor and to click and scroll. In Mount & Blade Warband, you can open the tactics menu with Backspace, and this lets you click on various troop types, orders, and on the minimap to hold positions.

Let mods add non-modal formspecs. Mobile users will be able to tap elements on them at any time, and PC users can push a key to release the mouse to the UI instead of camera controls. Now you don't even need touchscreen detection - but i you combine this with touchscreen detection, you can add touchscreen-specific controls. Thus, in an indirect way, we have what you asked for: buttons that can be used for controls on touchscreens.

There, we now have three separable items:
1. Bindings for keyboard (and other input devices like controllers) - suggested by #2731 and in issues linked from there - such as #8541. In fact the second suggestion in 8541 mentioned buttons just like you did - I still think non-modal HUD is the 'right implementation' of that basic idea.
2. Formspec/HUD hybrid that is non-modal - suggested [here](https://github.com/minetest/minetest/issues/10548)
3. Touchscreen gesture bindings - not suggested yet from a quick search.

No priority on any of those, not my call. Any objection to those ideas? Would having 1 & 2 satisfy your feature request here? Then my suggestion is the separate 3rd item.

**Comment by FatalError42O** [comment]
> Okay, let's say we can ignore gestures at the moment. What makes interactive tappable HUDs really a mobile-only feature? Plenty of desktop users want those too. Having formspecs be completely modal sucks, you can't see what's going on in the world behind the formspec.
> 
> Many games let you release the mouse from camera controls and interact with smaller UI elements. For instance in Garry's Mod, the spawn menu does not disable movement controls, but it redirects the mouse from camera controls to a cursor; the leaderboard can also be used with the mouse to mute players in voice chat. In Dark Souls, you can navigate the menus while continuing to move your character; some keys are bound to the menu, while the WASD movement keys continue to be useable, and the mouse can be used as a cursor and to click and scroll. In Mount & Blade Warband, you can open the tactics menu with Backspace, and this lets you click on various troop types, orders, and on the minimap to hold positions.
> 
> Let mods add non-modal formspecs. Mobile users will be able to tap elements on them at any time, and PC users can push a key to release the mouse to the UI instead of camera controls. Now you don't even need touchscreen detection - but i you combine this with touchscreen detection, you can add touchscreen-specific controls. Thus, in an indirect way, we have what you asked for: buttons that can be used for controls on touchscreens.
> 
> There, we now have three separable items:
> 
> 1. Bindings for keyboard (and other input devices like controllers) - suggested by [Adding custom keystrokes through Lua #2731](https://github.com/minetest/minetest/issues/2731) and in issues linked from there - such as [Control Scheme in Android #8541](https://github.com/minetest/minetest/issues/8541).
>    In fact the second suggestion in 8541 mentioned buttons just like you did - I still think non-modal HUD is the 'right implementation' of that basic idea, which brings me to..
> 2. Formspec/HUD hybrid that is non-modal - suggested in [Issues and suggestions regarding HUD elements #10548](https://github.com/minetest/minetest/issues/10548)
> 3. Touchscreen gesture bindings - not suggested yet from a quick search.
> 
> No priority on any of those, not my call. Any objection to those ideas? Would having 1 & 2 satisfy your feature request here? Then my suggestion is the separate 3rd item.

you're trying to solve problems that don't exist with this model for the sake of argument, and frankly I'm not willing to entertain your shit anymore. Making it a formspec, doesn't make sense at all, and you wanna hear the issue with this one chief? say you want to have a menu button, let's say.... X if you want mobile users to be able to use that, you have to have a button on the screen, that's not well scaled for PC on their screen? so now you both have a button on the screen, and a button that PC users don't want on their screen for no reason. And what if you have a gun mod, and you want to put a fire button for mobile people, now there's this annoying and useless button on the screen. 

Not to mention, I already proposed a solution for client detection not being implemented, it's very simple:

>If the inclusion of PR 12264 can not be made, HUD images with the button function should not be displayed on devices without capabilities to use them. 

it would be done entirely client side.

you're finding problems that don't exist, and arguing for the sake of arguing. I'm not going to entertain your shit anymore as thus far, everything you've said has been redundant. I'm not going to turn this whole comment section into a pissing competition.

**Comment by rubenwardy** [comment]
> - https://github.com/minetest/minetest/issues/2731

I knew this was a duplicate, but couldn't find the issue

**Comment by FatalError42O** [comment]
I appreciate that man

**Comment by Montandalar** [comment]
You didn't explain why making a formspec made no sense. It's basic code reuse in my opinion. Of course the formspec needs to be scaled, and of course it should only be shown to mobile users. You need to learn at least some agreeableness - you currently have displayed 0 tolerance for any other ideas. You can be as right as you want and still rude. Thank you for making sure I won't continue to contribute to this issue thread, because I don't want to deal with someone who isn't going to change their mind at all. I hope whoever next comments in this thread has a better time than me. You may as well start writing the PR yourself because if anybody disagrees with you about how to implement this and you do the same to me as to them, it won't turn out just the way you want it, and then you'll be mad your ideas were ignored or something.

**Comment by FatalError42O** [comment]
> You didn't explain why making a formspec made no sense. It's basic code reuse in my opinion. Of course the formspec needs to be scaled, and of course it should only be shown to mobile users. You need to learn at least some agreeableness - you currently have displayed 0 tolerance for any other ideas. You can be as right as you want and still rude. Thank you for making sure I won't continue to contribute to this issue thread, because I don't want to deal with someone who isn't going to change their mind at all. I hope whoever next comments in this thread has a better time than me. You may as well start writing the PR yourself because if anybody disagrees with you about how to implement this and you do the same to me as to them, it won't turn out just the way you want it, and then you'll be mad your ideas were ignored or something.

nope, you can't read.

>Making it a formspec, doesn't make sense at all, and you wanna hear the issue with this one chief? say you want to have a menu button, let's say.... X if you want mobile users to be able to use that, you have to have a button on the screen, that's not well scaled for PC on their screen? so now you both have a button on the screen, and a button that PC users don't want on their screen for no reason. And what if you have a gun mod, and you want to put a fire button for mobile people, now there's this annoying and useless button on the screen.

it's reimplementing a feature (i.e. regular formspecs) that already exists as a possibility within the fundamentals of the feature request (infact without modification of formspec code itself), AND it'd be broken and not make sense on PC. Furthermore, it doesn't solve more problems than it creates.

I have displayed 0 tolerance for YOUR ideas, because you clearly haven't thought this shit through, and are (once again as stated earlier) trying to have a pissing competition instead of make a acceptable and practical design.

I'm glad you won't contribute to this thread, infact I think it's great! Maybe you can even stop "contributing" to Minetest overall :D

**Comment by rubenwardy** [comment]
I don't really have time to read all of this right now, given the release, but I'd appreciate if this issue could get a little less heated as it doesn't seem healthy

**Comment by y5nw** [comment]
Sorry for seeing this thread this late. Here is my 2c on the matter:

> > Making it a formspec, doesn't make sense at all, and you wanna hear the issue with this one chief? say you want to have a menu button, let's say.... X if you want mobile users to be able to use that, you have to have a button on the screen, that's not well scaled for PC on their screen? so now you both have a button on the screen, and a button that PC users don't want on their screen for no reason. And what if you have a gun mod, and you want to put a fire button for mobile people, now there's this annoying and useless button on the screen.
> 
> it's reimplementing a feature (i.e. regular formspecs) that already exists as a possibility within the fundamentals of the feature request (infact without modification of formspec code itself), AND it'd be broken and not make sense on PC. Furthermore, it doesn't solve more problems than it creates. Also what do you mean it should only be shown on mobile, that directly conflicts with your own design philosophy for it-
> 
> > PC users can push a key to release the mouse to the UI instead of camera controls
> 
> It doesn't make sense for it to be a formspec, because formspecs are for FORMS, MENUS not controls, while HUD is not for buttons, it makes sense with how hud currently works to make this an addition.

I am seriously not sure what your arguments here are, especially in terms of why formspecs and HUD _must_ have different features and be mutually exclusive to each other.

At the current stage of Minetest, formspecs are the _only_ (sane) way to develop a UI more complex than something at a "hello world"-level, unless you want to go down the path of (1) writing abstractions for HUDs/user input handling, which I partly do, or (2) extending HUDs in such a way that it would likely duplicate formspec code anyway. Montandalar mentioned Advtrains HUD as an example, and, quite frankly, I would not be surprised if there were other mods with HUDs of that complexity.

HUDs and formspec essentially do the same thing - displaying stuff on the user's screen and interact with the user; what currently differentiates them are essentially arbitrary limitations that are made for some reason (I do not blame anyone for this, though, considering the scale of the project and the mere fact that people work on a voluntary basis), although I do not have enough knowledge of the UI code to go too far into the details.

Back to the original topic on supporting custom keybindings, I would suggest a hybrid of extending the HUD (or formspec) for mobile/touchscreens and using keybindings otherwise, as Minetest currently already does so.

**Comment by kneekoo** [comment]
Better input handing is very important for many use cases. And while not everything can be done the same way between platforms (PC vs mobile), one should not hold the other "hostage" towards offering a better experience. Ideally, each platform would be able to evolve separately, so that both platforms can evolve at the same time or at different paces without merge conflicts.

**Comment by appgurueu** [comment]
We have #14456 now, which should lay a good foundation for adding more customizable touchscreen controls. I'm thinking this could let us design an API roughly like this:

```lua
core.custom_control.register("mycontrol", {
    touch = {
        pos = {x = 0 ... 1, y = 0 ... 1}, -- default
        size = {x = 0 ... 1, y = 0 ... 1}, -- default
    },
    key = "a" ... "z" or "0" ... "9"
})
```

paired with something like a `player:get_custom_control("mycontrol")` or `player:get_custom_controls()`.

Additionally, there should be a `core.custom_control.enable` / `core.custom_control.disable` to show / hide these controls: For example a "reload" control should only be usable when a weapon is equipped and otherwise not clutter the precious mobile screen space.

On mobile, this would just add the touch controls, and make them accessible to the touch control editor so they can be rearranged (unconditionally, even if they are presently disabled).

On desktop it would add the control to the key change dialog and automatically configure it, unless there is a collision with an existing keybind or no corresponding scancode can be found for the given keycode, in which case it would remain unconfigured.

In both cases, the configured keybind / control position should be remembered.

Not all keys have good, universal keycodes ("names"), but some basic ones like letters, digits, arrow keys should work pretty much universally (and our docs should make clear that modders can not expect anything fancier to work generally). This is important so that keys like "R" (for reload) can just work out of the box, without players having to configure them.
Worst case we can't find the scancode for a given keycode and the user has to configure it themselves.

**Comment by grorp** [comment]
There's an old API proposal by rubenwardy for this here: https://github.com/minetest/minetest/issues/11446#issuecomment-883458358

**Comment by grorp** [comment]
For the touchscreen part:

In general, we need an API for games/mods to customize the touchscreen layout, not only to add custom buttons. Giving modders an API to place individual touchscreen buttons without giving them access to the predefined buttons means they'll have to rely on the default layout, which may change.

On the other hand, a small, scope-limited API like the one you proposed is still better than no API at all. We could add an API like this and deprecate it later in favor of a more comprehensive touchscreen controls API, or the more comprehensive API might never happen.

Now on the actual API: You're missing `texture` and `offset` fields. See also https://github.com/minetest/minetest/blob/ac7406c8a1013ea8bfc94c48dbc5dc97532c0cbd/src/gui/touchscreenlayout.h#L59

**Comment by y5nw** [comment]
(Nitpicking:) For keyboard controls, there are IMO some considerations to be made (especially with #14940) regarding the implementation:
* An API for getting players' keyboard configurations should be able to get both the scancode and the keycode, as both have their usecases in terms of showing the control scheme to the user. (That said, this particular usecase should ideally be implemented with SSCSMs instead of server-sent UIs. Alternatively/Additionally, an escape sequence can be used as a placeholder for the key name.)
* Getting the (user facing) key name would require additional changes to the translation system depending on the implementation, as the key names are currently translated on the client-side using gettext instead of using the translation system provided by the engine. (A more general solution would be to move client UI strings to use the translation system provided by the engine, but this would be out of scope for a PR implementing custom keybindings.)
* There should be a way to specify alternative/secondary keybindings:
  * There will very likely be collisions when multiple mods define their own keybindings, or
  * For symmetry (e.g. allow using Left or Right Control for an action); related: https://github.com/minetest/minetest/pull/14964#issuecomment-2335193253

For touchscreens controls, I recall reading somewhere (possibly from resolving merge conflicts) that touchscreen input internally looks for a corresponding keycode/scancode and (only if such a keybinding is present) sends the corresponding keypress event. I am not sure whether this remains the case, but this will need to be addressed as well if it is.

**Comment by pkleineb** [comment]
I would love the feature proposed here. Is there any plan to implementing this in comming releases of luanti?

**Comment by rubenwardy** [comment]
https://github.com/luanti-org/luanti/pull/16053#issuecomment-2857730778

**Comment by eligarner23** [comment]
ok i was about to suggets a simialr issue but for the special1 keybind to be accessible as well

**Comment by y5nw** [comment]
> ok i was about to suggets a simialr issue but for the special1 keybind to be accessible as well

I do not think `special1` exists at all as a keybinding.

Not to mention that exposing random hardcoded keybindings to the server modding API is an XY problem to begin with, and this issue is about the more fundamental limitation.

**Comment by eligarner23** [comment]
ok thanks for the cofirmation, is there any plans for what this feature request asked for in the next releases of Minetest?

**Comment by y5nw** [comment]
> is there any plans for what this feature request asked for in the next releases of Minetest?

This would depend on your expectations on a "plan". Prior work has largely focused on the prerequisites for implementing this feature properly (SSCSMs and reworking the input handler, among other things), and these still need to be worked on (see roadmap). The amount of effort put _specifically_ into this feature request is (AFAIK) largely limited to ideas on how such an API would look like, and (IMO) it currently does not make much sense to consider further implementation details given that non-trivial changes to the (architecture of the) input handler are also expected in the medium term (I do not believe I have sufficient knowledge to assess the current state of SSCSMs with regard to this feature).
